# On developing
All members add note to this thread by (Name - (id:your_favoriate_number_with_four_numbers))

- Nam (id:1321): The image file (.jpeg) is the arts I created for this course. The name seems strange and I will explain them later during the course. Of note, these assets **are not** crypto-currency. This encoding is quite common if you using Unity, they have a special way to encode those assets. Before using of any assets created by this notebook, we must accept these following policy:

All actions extended from this notebook that mean harm for human beings and lives are prohibited.

Definition of **Harm** includes
- Hate: hateful symbols, negative stereotypes, comparing certain groups to animals/objects, or otherwise expressing or promoting hate based on identity.
- Harassment: mocking, threatening, or bullying an individual.
- Violence: violent acts and the suffering or humiliation of others.
- Self-harm: suicide, cutting, eating disorders, and other attempts at harming oneself.
- Sexual: nudity, sexual acts, sexual services, or content otherwise meant to arouse sexual excitement.
- Shocking: bodily fluids, obscene gestures, or other profane subjects that may shock or disgust. Illegal activity: drug use, theft, vandalism, and other illegal activities.
- Deception: major conspiracies or events related to major ongoing geopolitical events.
- Political: politicians, ballot-boxes, protests, or other content that may be used to influence the political process or to campaign.
- Public and personal health: the treatment, prevention, diagnosis, or transmission of diseases, or people experiencing health ailments.
- Spam: unsolicited bulk content.
- **Monetize: All resources generated within computation/simulation cannot be transferred to actual money, asset in any circumstances. Attempt to violate this policy will cope with legal issues.**

# Prolouge
(Author: Nam Nguyen)

Course Introduction:
This is a two-semester workshop course on AI, Quantum ML and CRISPR Design. By the end of the workshop, we will:
- Learn how to generalized modern ML via Abstract Algebra Language
- Learn many cool hand-ons experiments
- Learn the TEEA (*The Theory of Every Species, Everywhere in the Universe, All at one superposition of Quantum Systems*)

About Me:
I am Nam Phuong Nguyen (Nguyễn Phương Nam), and I am from Vietnam. I have a special inclination for teaching mathematics as I have been a highschool math teacher for a while before my great venture to the US. These lecture notes will be my first attempt to make you understand my theory of life science, the TEEA.

I want to thank my dear parents. Even though we lived miles away during the completion of this thesis, they are always by my side. This thesis represents the Hilbert ($\mathcal{H}$) space of Quantum Regime and the sample space $\Omega$ of probability theory. My brother ($\mathbb{P}$) for a probability density function and my fiance $\aleph_0$ - the natural numbers...

# Artificial Intelligence from Linear Algebra
(Author: Nam Nguyen)
## Theory
Machine Learning, in the most simple explanation, is a computational model that can (1) learn from data, (2) analyze those data and (3) make a prediction based on empirical evidence. Deep Learning is a special type of Machine Learning in which the model includes a stack of layers with neurons. We can imagine a computational neuron as a switch, presented as a step function taking binary values of 0 and 1. In Figure 1, we can imagine that the more neurons included in the model, the more step function we can use to estimate the true function, colored in (*a*).

Now, let us go through three complexities of ML problems, including (1) Scalars, (2) Vector, and (3) Vector Fields.

### Scalar
An example of the scalar is the real number we use every day, for which a scalar of currency $3.14$ can be mapped to a straight line slope in 2 dimensions $y=3.14x$. Machine Learning on a single scalar field is the most simple case. Suppose you open a vendor to sell a type of product $A$, with $3.14$ for each item. So, the total venue that you gain is $y = 3.14x$ for $x$ is the number of items you sold and $y:=$ total venue for each day. If you have a notebook and formulate a table for $y$, you will have a random variable of $y$, denoted in a more fancy way as $Y$.

Why is $Y$ a Random Variable? You are not a superficial character, so you can never predict precisely the number of the item that could be sold daily. However, if you keep track of your sale in a great amount of time, you could have a model to predict your future outcome - the Machine Learning model. Suppose the dataset you collect in the (*tabular*) form:
| Day         |(# Items)            | Total ($y$)  |
| ------------- |:-------------:| -----:|
| 1 | 11|$ |
| 3 | 23|$ |
| 5 | 19|$ |
| 7 | 17|$ |
| 11 | 29|$ |

Now, let us build a simple model for a random variable $Y$ in Python:


In [1]:
# Declare your package
import numpy as np #Use for Array Manipulation, Transformation
import pandas as pd #Use for DataFrame Process

day = np.array([1,3,5,7,11]).astype('int') # The Integer
n_item = np.array([11,23,19,17,29]).astype('float') #The Real Number

def y_sale(n_item, nest_price = 3.14):
    '''
    y_sale: is the random varible that we want to compute in the targeted column
    '''
    y = n_item*3.14
    return y

print(y_sale(n_item))

[34.54 72.22 59.66 53.38 91.06]


So, the function we create is a Simple Linear Regression model of random variable $Y$ on the number of item sold daily $x$, which as a tuple
$$Y = \{34.54, 72.22, 59.66, 53.38, 91.06\}$$ 

Indeed this is a very limited data-scale, but we can compute some important statistical quantities such as mean or standard variation simply by:

In [2]:
y = y_sale(n_item)
y_mean = y.mean()
y_std = y.std()
print('The Mean Sales: {}'.format(y_mean))
print('The Standard Deviation of Sales: {}'.format(y_std))

The Mean Sales: 62.172000000000004
The Standard Deviation of Sales: 18.881820251236373


The mean and variance of a random variable can be expressed as:

[**Mean**]
The mean of a random variable X is defined as:
\begin{equation} 
\mu = E[X] = \int_{-\infty}^{\infty} x f(x) dx
\end{equation}
where $\mu$ is the mean of X, $E[X]$ denotes the expected value of X, and $f(x)$ is the probability density function (PDF) of X.

[**Variance**]
The variance of a random variable X is defined as:

\begin{equation} 
\sigma^2 = Var[X] = E[(X - \mu)^2] = \int_{-\infty}^{\infty} (x - \mu)^2 f(x) dx
\end{equation}

where $\sigma^2$ is the variance of X, $Var[X]$ denotes the variance of X, $\mu$ is the mean of X, and $f(x)$ is the probability density function (PDF) of X.

We will demystify these concepts in (**Activity 4: The Duality Game**).

Now, let's try different scenarios here:
- [A0] You're a Biochemist.
- [A1] You're a Mathematician
- [A2] You're Z.E.N.U.S
![Agent Selection](TEEA_v0.005.jpeg)

(**The Biochemist**)
He works with metal all the time. Today is the first day he found Zn. He opens up his G.A.I.A and take a note:
![kem](TEEA_v0.012.jpeg)


He builds a Decison Tree on top of the Atomic Mass of Zn, which is $65.38$. The Decisison is given as:
\begin{equation}
    y_{\text{new}} = \text{action: raise price to 4.13}, \text{if } E[Y] > A_{\text{Zn} = 65.38};
\end{equation}
otherwise keep the same price.

The simple Decision Model is programmed as follow:

In [3]:
def OneDecisionTree():
    if y_mean > 65.38:
        print('Raised Price:')
        y_new = y_sale(n_item, 4.13)
    else:
        print('Keep Price:')
        y_new = y_sale(n_item)
    return y_new

print(OneDecisionTree())



Keep Price:
[34.54 72.22 59.66 53.38 91.06]


![Reward](TEEA_v0.022.jpeg)
![Ice](i71838b53228ad6332612892564d602bb9d25a36a218501921c66f66591465f7.001.jpeg)
![Fire](f71838b53228ad6332612892564d602bb9d25a36a218501921c66f66591465f7.001.jpeg)


### Vector
Now, all of you will either have Ice or Fire. This is because our simple Decision Tree has only two outcomes: (1) Raised price if Mean Sales greater than Atomic Mass of Zn, otherwise (2) Keep the same price. Let's try the second senerio:

(**Fire**) @-235711 is a package on the train that Fire wants to steal. She must make her molecular vibrating as wave to infiltrate the train. She opens her G.A.I.A, looking at the Bohr and sparking a thought: "Oh, yeah 3.14!".
In this senarior, there demands a non-linear relation. Of note, the function we have in the first senerio is linear (recall: $y = a\times x, a = 3.14$). We can construct a complicated function by adding more term into the function, such as adding (**bias term**), or intersection with $Oy$. Specifically, the model is now more generalized given as
\begin{equation}
    Y = a\times X + b(\text{ias}).
\end{equation}

Now, we just adding one more dimension into the sale function. By no mean this is a geometric dimension as we still see the function as a 2D line in the case of $Y = ax$. The dimensional extension here happens at the functional basis of polinomial classes. Specifically, the linear regression model $Y = ax$ is spanned by (**polynomial term**) $\{x\}$. Fire must added more terms to her function such as 
\begin{equation}
    Y = 1+z + z^2 +z^3,
\end{equation}
here we present the equation for Fire as variable $z$ to distinguish from the Biochemist scene.

So, this is Linear or Non-Linear?
- Linear (Ice)
- Non-Linear (Fire)
- ..?! (S)

[**Fire**] says: Let's non-linear..:>>

In [4]:
import plotly.express as px

def f_non_linear(x):
    y = 1 + x + x**2 + x**3
    return y
z = y_sale(n_item)
y_z = f_non_linear(z)
fig = px.scatter(z,y_z)
fig.update_layout(xaxis_title = '<b>Y</b>')
fig.update_layout(yaxis_title = 'Non-Linear Thoughts (<b>f(Y)</b>)')
fig.show()

[**Ice**] says Linear. Ice was right, the statistical function $$Y = 1 + z + z^2 + z^3$$ is a linear model in term of statistical learning. Specifically, the functional basis now has four element $\{1,z,z^2,z^3 \}$. The construction of dependent (predicted) variable $Y$ is a linear combination of such a functional basis with coefficient $(1,1,1,1)$. In general, we can present a polynomial model as the linear combination 
\begin{equation}
    h(z) = a_0 z^{0} + a_1 z^{1} + a_2 z^{2} + \dots + a_k z^{k} = \sum_{i=1}^{k} a_i z^{i},
\end{equation}
$\alpha_i \in \mathbb{R}$ (we says $alpha_i$ in real-valued vector space).

Fire did not make is this time as she reach 763K. Let's start again.

### Vector Field
You're Galois, but now people known you as SINDy (Let's dicuss the othertime). The Prime Fields is a very special concept in mathematics. Let's pick a favorite, natural number, you will have two senarior: prime or non-prime. A prime number is a positive integer greater than 1 that has no positive integer divisors other than 1 and itself. In other words, a prime number is only divisible by 1 and itself. For example, the first few prime numbers are 2, 3, 5, 7, 11, 13, 17, 19, 23, and so on.

[**Fire**] wakes up. She dreams about a strange dream last night. She looks at her cell, where she hangs (**the Star Aligments**)
![Star Aligment](TEEA_v0.019.jpeg)
She was late for Dr. Galoises's class.

She has no-time to teleport to all these stars to reach the Academy. So, she decide to go only for a Prime Stars.

Definition: A Prime Star is encoded as follow:
- s orbital: can hold up to 2 electrons
- p orbital: can hold up to 6 electrons
- d orbital: can hold up to 10 electrons
- f orbital: can hold up to 14 electrons
with prime number of electrons.

Now, [**Fire**] will have much bigger problem than the Biochemist ($Y = 3.14 X$) and the one she dreamt ($f(z) = 1+z+z^2+z^3$), ploted in her thought (Figure: Non-Linear Thoughts). Now, she must create a prime map from (*her poster on the shelf..?!*)



In [5]:
shelf = ['s','p','d','f']
n_ele_shelf = [2,6,10,14]

electron_dict = dict(zip(shelf,n_ele_shelf))


gstr = ['1s', '2s', '2p', '3s', '3p', '4s', '3d', '4p', '5s', '4d', '5p', '6s', '4f', '5d', '6p', '7s']
estr = ''.join(gstr)
print(estr)

n_s, n_p, n_d, n_f = [], [], [], []
for e in estr:

    if e == 's':
        n_s.append(electron_dict[e])
    elif e == 'p':
        n_p.append(electron_dict[e])
    elif e == 'd':
        n_d.append(electron_dict[e])
    elif e == 'f':
        n_f.append(electron_dict[e])
    else:
        pass
print(n_s)
print(n_p)
print(n_d)
print(n_f)

1s2s2p3s3p4s3d4p5s4d5p6s4f5d6p7s
[2, 2, 2, 2, 2, 2, 2]
[6, 6, 6, 6, 6]
[10, 10, 10]
[14]


Let's have a quick check. We have (**s**)even orbitals, with 1s2s3s4s5s6s7s; five p, three p and a f..?! If we construct a function
\begin{equation}
    \eta = \alpha_0 s + \alpha_1 p + \alpha_2 d +\alpha_3 f
\end{equation}
in which we have $\alpha_i:=$ counting the number of characters $\{s, p, d, f \}$. With the count from string collected from Fire poster, we have $\alpha_0 = 7, \alpha_1 = 5, \alpha_2 = 3, \alpha_3 = 1$.

Now, if we define $s = x^0$, $p = x^3$, $d = x^5$ and $f = x^7$, we have the eta function presented on $\{1,x^3,x^5,x^7\}$ basis, much complicated than the one she dreamt last night $f(z) = 1+z+z^2+z^3$.

(*Definition*)
A vector field is a function that assigns a vector to each point in a given space, and can be expressed mathematically as:

$$\textbf{F} : \mathbb{R}^n \rightarrow \mathbb{R}^n$$

where $\textbf{F}$ is the vector field, $\mathbb{R}^n$ is the n-dimensional (**Euclidean space**), and $\rightarrow$ denotes a mapping from one space to another. The vector at each point $x \in \mathbb{R}^n$ is given by:

$$\textbf{F}(x) = \begin{bmatrix} F_1(x_1, x_2, ..., x_n) \ F_2(x_1, x_2, ..., x_n) \ \vdots \ F_n(x_1, x_2, ..., x_n) \end{bmatrix}$$
where $F_1, F_2, ..., F_n$ are the scalar functions that determine the components of the vector field.

We will discuss the concept of Prime Fields in (**Activity 8: Quantum Encoding of Languages**).



When the computation done, Fire opens her G.A.I.A, putting the three elementaries into the Forge:
![theKIN](TEEA_v0.036.jpeg)

#### Brief Historical Notes
[ChatGPT]: Évariste Galois was a French mathematician born in 1811 who made significant contributions to the field of algebra, particularly in the study of algebraic equations and their solutions. He is known for his work in Galois theory, which explores the relationship between field theory and group theory, and provides a method for determining whether an algebraic equation can be solved using radicals.

Galois theory establishes a correspondence between field extensions and groups, which allows one to understand the structure of the field extension by studying the corresponding group. Galois also introduced the concept of a normal subgroup, which is essential in the study of groups and their properties.

Galois died at the age of 20 in a duel, but his work was later recognized and expanded upon by other mathematicians. His contributions to algebra and group theory have had a profound impact on mathematics and other fields, such as physics and cryptography. The Galois group, Galois fields, and Galois connections are just a few examples of the many mathematical concepts that are named after him.


## Data Structure for Modern AI
Galois looks at the green skies, beside his (agile) deer. Today is the first field trip to him as his team has to pass through the **Force** of data structure. We leave our digital envidence whenever you goes, in all (**data structure form**). Generaly speaking,
- Our physical system are $3D$ object, with one unseen dimension of time (Discuss Later).
- Our photos are $2D$ array with three channels (color): $X^{\text{width}\times \text{height} \times \text{dim channel=3}}$ or one channel gray. 
- Our text are sequences, in $1D$ array with mathematical expresion $X_{\text{n sequence}} = NATGC$

[ChatGPT]: Scientists studying the genetic code of organisms have identified four types of nitrogenous bases, which are abbreviated as NATGC, and these bases pair up in a specific way to form the building blocks of DNA.



## The Zoo of Architectures
So, for each type of data, we will have a corresponding type of AI models for each data types. The term "Zoo fo Model Architectures" was coined in the concept of Geometric Deep Learning[Cite]. Let's demonstrate the concepts as follow:
- Take a look at the variable: "estr" = 1s2s2p3s3p4s3d4p5s4d5p6s4f5d6p7s..?!
- Take a look at the picture of Galois and his deer..!?
- Take a look at you, now, and then!










Galois can select a partner for his field trip.
[Cliche Quote: "*The power of teamwork lies in the collective efforts of individuals working together towards a common goal, combining their unique strengths and abilities to achieve a shared vision that may have been impossible to accomplish alone. When people work as a team, they can often accomplish more, make better decisions, and be more productive than when working individually. Moreover, teamwork can foster a sense of camaraderie and support, allowing team members to learn from one another, share ideas, and grow both personally and professionally. Overall, the power of teamwork can be transformative, enabling individuals to accomplish great things and create positive change in their lives and communities...*"]

Let's start again. Which one you choose, Fire?
So, this is Linear or Non-Linear?
- Linear (Ice)
- Non-Linear (Fire)
- ..?! (S)

Yeah. Here me right, reader. You're Fire.
![scence_2](TEEA_v0.005.jpeg)

[**ICE**]
Suppose Galois and His Friend go to the Forge and meet a creature that never see before, with the data collected as

In [6]:
print(estr)

1s2s2p3s3p4s3d4p5s4d5p6s4f5d6p7s


He's already has a Simple Linear Model Eta given as
\begin{equation}
    \eta = \alpha_0 s + \alpha_1 p + \alpha_2 d +\alpha_3 f.
\end{equation}
with hypothetical envidence as count statistics for each random variable $\{s,p,d,f \}$.

In [7]:
print(n_s)
print(n_p)
print(n_d)
print(n_f)

electron_shield = [n_s, n_p, n_d, n_f]
electron_shield = [np.array(x).sum() for x in electron_shield]
print(electron_shield)

[2, 2, 2, 2, 2, 2, 2]
[6, 6, 6, 6, 6]
[10, 10, 10]
[14]
[14, 30, 30, 14]


Both of model $\eta$ and the input genetic code of this creature are linear. Let's discuss another scenario.

[**fire**] A non-linear (*representation*) of the genetic code. This is a fictional example to demonstrate a toy model.

We can construct a numerical tensor, or matrix of the form
$$\begin{pmatrix}
\sum_{ns} & \sum_{np}\\
\sum_{nd} & \sum_{nf}
\end{pmatrix}_{2\times 2}$$

In [8]:
sigma_electron = np.array([electron_shield[0],electron_shield[1],
                           electron_shield[2],electron_shield[2]])
sigma_electron = sigma_electron.reshape(2,2)
print(sigma_electron)
fig = px.imshow(sigma_electron)
fig.show()


[[14 30]
 [30 30]]


## Decide Neural Architecture for Your Models
Now, let's practice with this scenarios. Given that you collect a bunch of genetic codes, given in these data collections:
- CRISPR Base: Link
- DS-CRISPR: Link

## QUEST:
Given the input string $g_{1\times m}$, with $m$ is the length of the string code for the input string (CRISPR-Based) and (DS-CRISPR). We have three scenarios for sgRNAs, with the nucleotide (nt) length of:
- 26 nt
- 30 nt
- 32 nt

In the (**stat_meta_**), we have given the motif counts of the genetic string. 

# Hand-ons:
- (**Step 1**): Understand the Dataset: We are given several features. Now, let's summarize them.



In [9]:
data = pd.read_csv('stat_meta_GFDS_26.csv')
print(data)
print(data.columns)

        DS-Seq                       sgRNA  \
0         1234  TAGTTACTGAGAGATTAGTACGTTCA   
1         1234  GCACTACGAGCGACGGTATGTTCAAC   
2         1234  GTGCGCCGACAGCTGTCGTGCTTCCT   
3         1234  GTCGTCCGATCCTTAATCAGATGTCG   
4         1234  GCGCACTCAGCCTCACACCAGGCAGA   
...        ...                         ...   
293571    4321  GTGCAGGGTACGGTCAGAGTTAGTAT   
293572    4321  GCCTGCATTAGCACAGCAGCGTCTGC   
293573    4321  CGTCGTGTAAGTGTCCCAGTCGTCGT   
293574    4321  TACGCATCGTAGCTGATACCTATCAC   
293575    4321  GTTGAGCAGAGACCACTACCAGTATC   

                                                     bits  Global_GC  \
0       0100110101001001110011001100010100110100101101...   0.346154   
1       1110001001001011001110110010111101000111010110...   0.538462   
2       1101111011101011001000111001110110110111100101...   0.653846   
3       1101101101101011000110100101000001100011000111...   0.500000   
4       1110111000100110001110100110001000101000111110...   0.653846   
...            

Suppose we only have the knowledge that has been established up to this point. We can seperate the features into two sets:
- Unknown: DS-Seq, bits, sA, sT, sG, sC
- Known: Motif (counts of A, AG, CGG, pAGG, etc). p is representing the Protospace Adjacency Motif (PAM).

We will discuss this further in (**Activity 6: Send and Envelop to the Human Genome**).

- (**Step 2**): Define Objectives
We want to use the string code to predict one of the variable, suppose we select GC-content. Now, we have the model
$$\hat{y}_{\text{GC}} = f(\text{sgRNA})$$
We use the hat notation for the reason that $\hat{y}$ is only estimation of the actual GC-content based on analyzing the input string sgRNA.

In [10]:
import os
from sklearn.model_selection import train_test_split
import xgboost as xgb
import optuna
from scipy import stats
import math
from optuna.samplers import TPESampler

# ENCODING METHOD
def OneHotEncoding(x):
    X = []
    for i in range(len(x)):
        if x[i] == 'A':
            X.append(np.array([1,0,0,0]))
        elif x[i] == 'T':
            X.append(np.array([0,1,0,0]))
        elif x[i] == 'G':
            X.append(np.array([0,0,1,0]))
        elif x[i] == 'C':
            X.append(np.array([0,0,0,1]))
        else:
            print('Invalid RNA sequences')
    X = np.concatenate(X, axis = 0)
    return X

def train(sgRNA, scores, dataname,
    feature_name, feature_dict, n_trials, 
    nepochs, seed, _savedir):
  savedir = 'ExpHist_{}_{}_{}_{}'.format(_savedir,dataname,feature_name, seed)
  modeldir = os.path.join(savedir,'save_models')
  try:
    os.mkdir(savedir)
    os.mkdir(modeldir)
  except:
    pass
  train_idx, test_idx = train_test_split(range(len(sgRNA)), random_state = seed)
  print(feature_dict[feature_name])
  print(train_idx)
  x_train = np.array(feature_dict[feature_name])[train_idx]
  y_train = scores[train_idx]
  x_test = np.array(feature_dict[feature_name])[test_idx]
  y_test = scores[test_idx]
  print(x_train.shape)
  print(x_test.shape)
  print(y_train.shape)
  print(y_test.shape)
  # LOAD DATA FOR xgBoost
  dtrain = xgb.DMatrix(x_train, label=y_train)
  dtest = xgb.DMatrix(x_test, label=y_test)
  evallist = [(dtrain, 'train'), (dtest, 'eval')]
  def objective(trial):
  # CREATE HYPER-PARAMETER SEARCH SPACE
    booster = trial.suggest_categorical('booster', ['gblinear', 'dart'])
    eta = trial.suggest_float('eta', 0, 1)
    max_depth = trial.suggest_int('max_depth', 1, 11)
    min_child_weight = trial.suggest_float('min_child_weight', 0, 11)
    max_delta_step = trial.suggest_float('max_delta_step', 1, 11)
    subsample = trial.suggest_float('subsample', 0, 1)
    _lambda = trial.suggest_float('lambda',0 , 5)
    alpha = trial.suggest_float('alpha',0, 5)
    grow_policy = trial.suggest_categorical('grow_policy', ['depthwise', 'lossguide'])
    param = {'booster': booster,
             'eta': eta, 
            'max_depth': max_depth,
            'min_child_weight': min_child_weight,
            'max_delta_step': max_delta_step,
            'subsample': subsample,
            'lambda': _lambda,
            'alpha': alpha,
            'grow_policy': grow_policy,
            'objective': 'reg:squarederror'}
    param['nthread'] = 12
    param['eval_metric'] = 'rmse'
    # TRAIN
    bst = xgb.train(param, dtrain, nepochs)
    y_test_pred = bst.predict(dtest)
    score = stats.spearmanr(y_test_pred, y_test, nan_policy='omit')[0]
    if math.isnan(score) == True:
      score = 0
    bst.save_model('{}/xgboost_{}__{}.json'.format(modeldir, str(trial.number+1).zfill(3), score))
    return score
  # SAVE EXP
  np.save('{}/train_idx.npy'.format(savedir), train_idx)
  np.save('{}/test_idx.npy'.format(savedir), test_idx) 
  # OPTIMIZE 
  study = optuna.create_study(sampler=TPESampler(seed = seed),study_name = savedir,
                              storage = 'sqlite:///{}/{}.db'.format(savedir,savedir),
                              load_if_exists=True,
                              direction = 'maximize')
  study.optimize(objective, n_trials=n_trials)
  return True

KeyboardInterrupt: 

In [ ]:
# SUB-1000
data = data[:1000]
gstr = data['sgRNA']
gc_scores = data['Global_GC']
feature_dict = {}
feature_dict['OHE'] = [OneHotEncoding(g) for g in gstr]
# train argument:
# sgRNA, scores, dataname,
# feature_name, feature_dict, n_trials, 
# nepochs, seed, _savedir

train(sgRNA = gstr, scores = gc_scores, dataname = 'GFDS_26_sub1000', feature_name='OHE',feature_dict=feature_dict, 
      n_trials=1,nepochs=100, seed = 1,_savedir = 'activity_1_gfds_26_sub1000')

[I 2023-04-09 17:49:29,674] Using an existing study with name 'ExpHist_activity_1_gfds_26_sub1000_GFDS_26_sub1000_OHE_1' instead of creating a new one.


[array([0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0]), array([0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1,
       0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1]), array([0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0,
       0,

/Users/namnguyen/opt/anaconda3/envs/quanonco/lib/python3.7/site-packages/scipy/stats/stats.py:4484: SpearmanRConstantInputWarning:

An input array is constant; the correlation coefficient is not defined.

[I 2023-04-09 17:49:30,056] Trial 51 finished with value: 0.0 and parameters: {'booster': 'gblinear', 'eta': 0.9303052323857038, 'max_depth': 2, 'min_child_weight': 9.624818284697817, 'max_delta_step': 2.4370032517093905, 'subsample': 0.9408871896415086, 'lambda': 1.2075466850463654, 'alpha': 0.9262171734116129, 'grow_policy': 'depthwise'}. Best is trial 49 with value: 0.8913905610253968.


[17:49:29] WARNING: /Users/runner/work/xgboost/xgboost/python-package/build/temp.macosx-10.9-x86_64-3.7/xgboost/src/learner.cc:627: 
Parameters: { "grow_policy", "max_delta_step", "max_depth", "min_child_weight", "subsample" } might not be used.

  This could be a false alarm, with some parameters getting used by language bindings but
  then being mistakenly passed down to XGBoost core, or some parameter actually being used
  but getting flagged wrongly here. Please open an issue if you find any such cases.




True

# Key takes
To the end of this activity, we have learn:
- Build a Simple Linear Regression and a Simple Decisio Tree based on Automic Mass.
- Distinguish between Non-Linear Dynamics and Statistical Linear Model (Linear combination of polynomial terms).
- Data structure for computational models.
- Buil the first xgBoost (Gradient Boosted Tree) for an example datasets.

# GPT-guided concepts:
- Integral Domain, Group and Lie Algebra.